### PR TITLE
added get_extension_for_class #2255

### PR DESCRIPTION
**Pull request by queenp**
Note: I've used "for", rather than get_extension_by_class, matching the original "get_extension_for_oid".

### CODE REVIEW

**Comment by reaperhulk** [comment]
Looks like you need to rebase this against current master to resolve some conflicts (we've been moving things around recently)

**Comment by queenp** [comment]
Think it's sorted now.

File: src/cryptography/x509/extensions.py [hunk 1]
```diff
@@ -80,6 +80,13 @@ def get_extension_for_oid(self, oid):
 
         raise ExtensionNotFound("No {0} extension was found".format(oid), oid)
 
+    def get_extension_for_class(self, extclass):
+        for ext in self:
+            if type(ext.value) == extclass:
```
Comment (reaperhulk) [inline src/cryptography/x509/extensions.py]: This should be done with isinstance rather than type.

**Comment by reaperhulk** [comment]
You'll need to add a test that uses this method. We require 100% code coverage.

Thanks for your work so far!

**Comment by queenp** [comment]
Sure thing, would you recommend that I mirror the full range of get_extension_for_oid checks?

**Comment by reaperhulk** [comment]
In this case a single test that loads a cert, then uses the method to get the extension (and asserts on the result) should be sufficient.

**Comment by reaperhulk** [comment]
Looks like to get full coverage you'll need to add a test that raises `ExtensionNotFound`.

**Comment by reaperhulk** [comment]
jenkins, ok to test

**Comment by queenp** [comment]
Sorry for the messy commits. First time contributing to a strictly quality controlled repo. Bit of a learning curve.

**Comment by queenp** [comment]
I'm completely at a loss wrt what the branching condition is. My function and tests are almost identical to get_extension_for_oid. I can't find any condition that is left untested unless it's a quirk of isinstance()

File: docs/x509/reference.rst [hunk 2]
```diff
@@ -889,6 +889,21 @@ X.509 Extensions
             >>> cert.extensions.get_extension_for_oid(ExtensionOID.BASIC_CONSTRAINTS)
             <Extension(oid=<ObjectIdentifier(oid=2.5.29.19, name=basicConstraints)>, critical=True, value=<BasicConstraints(ca=True, path_length=None)>)>
 
+    .. method:: get_extension_for_class(extclass)
+
```
Comment (reaperhulk) [inline docs/x509/reference.rst]: Could you add a `.. versionadded:: 1.1` here?

**Comment by queenp** [comment]
Okay, I added that version tag to the doc page and literally didn't touch the test code or the function and code coverage panicked on the "isinstance" line again.

**Comment by reaperhulk** [comment]
Weird... For the case where it doesn't find the extension try switching to using `x509/cryptography.io.pem` and have it look for `IssuerAlternativeName` (since BasicConstraints are actually present in the cryptography.io.pem cert)

**Comment by queenp** [comment]
Sure, but that's not what's failing, so far as I can tell. It intermittently fails and passes the "if isinstance()" line (I've tried to eliminate any problems by making it as close to the branching logic of get_extension_for_oid() as possible).

**Comment by reaperhulk** [comment]
Thanks @queenp! If you'd like to have your name in the AUTHORS file feel free to send a PR :)

**Pull request merged**